src/IssueViz.Security/Taint/TaintStore.cs

```diff
@@ -39,21 +40,85 @@ internal interface ITaintStore : IIssuesStore
         /// Returns additional analysis information for the existing visualizations in the store.
         /// </summary>
         AnalysisInformation GetAnalysisInformation();
+
+        /// <summary>
+        /// Add the given issue to the existing list of visualizations.
+        /// If <see cref="GetAnalysisInformation"/> is null, the operation is ignored.
+        /// </summary>
+        void Add(IAnalysisIssueVisualization issueVisualization);
+
+        /// <summary>
+        /// Removes an issue with the given key from the existing list of visualizations.
+        /// If no matching issue is found, the operation is ignored.
+        /// </summary>
+        void Remove(string issueKey);
     }
 
     [Export(typeof(ITaintStore))]
     [Export(typeof(IIssuesStore))]
     [PartCreationPolicy(CreationPolicy.Shared)]
     internal sealed class TaintStore : ITaintStore
     {
-        private IAnalysisIssueVisualization[] taintVulnerabilities = Array.Empty<IAnalysisIssueVisualization>();
+        public event EventHandler<IssuesChangedEventArgs> IssuesChanged;
+
+        private static readonly object Locker = new object();
+
+        private List<IAnalysisIssueVisualization> taintVulnerabilities = Array.Empty<IAnalysisIssueVisualization>().ToList();
         private AnalysisInformation analysisInformation;
 
-        public IReadOnlyCollection<IAnalysisIssueVisualization> GetAll() => taintVulnerabilities;
-        
+        public IReadOnlyCollection<IAnalysisIssueVisualization> GetAll()
+        {
+            lock (Locker)
+            {
+                return taintVulnerabilities;
+            }
+        }
+
         public AnalysisInformation GetAnalysisInformation() => analysisInformation;
 
-        public event EventHandler<IssuesChangedEventArgs> IssuesChanged;
+        public void Add(IAnalysisIssueVisualization issueVisualization)
+        {
+            if (issueVisualization == null)
+            {
+                throw new ArgumentNullException(nameof(issueVisualization));
+            }
+
+            if (analysisInformation == null)
```

Comment:
I think there are a couple of possible problems here.

It looks like the `TaintSynchronizer` will clear the store when a standalone solution is opened, and fetch taint issues when a connected solution is opened/updated. Until the fetch has been completed successfully, `analysisInformation` will be null.

A) possible race condition: we get a taint event notification before the `TaintSynchronizer` has fetched the initial set of taint issues -> `analysisInformation` will still be null -> event will be lost.

B) the `TaintSynchronizer` has finished its initial sync, but there are no taint issues to start with, so `analysisInformation` never gets set. A taint issue is added to the server -> taint event -> `analysisInformation` is `null` -> event is ignored.

I think (A) is a corner case that we can probably ignore. Hopefully the ignored event would be for a taint issue that would be in the batch fetch anyway.

(B) is a problem, but should be easy enough to fix.

Reply:
Hmm, B is indeed a problem. `analysisInformation` is used in the UI to display what branch we're on and when was the last analysis. Based on information we have in the event, we can know the branch name, but we don't know when the analysis happened. Perhaps we can use DateTime.Now? 

IMO the root cause here is that we want the list to be empty when there are no issues for the given file, so we want to distinguish between "there are no issues, at all" (= no analysis info), and "there are no issues in this file" (=has analysis info). I'm not sure how to tackle this.

Reply:
If we ignore the possible race condition in (A) (which I think is ok), we just need to change the `SynchronizeWithServer` method so that it always calls `GetAnalysisInformation(...)`, even if there are currently no issues.

There's a separate, unrelated issue: the taint synchronizer isn't branch-aware. We missed this case when doing the branch-awareness work :-(
I've created #3683.

Reply:
@duncanp-sonar it looks like taint itself _is_ branch aware, but the UI is displaying wrong information. https://github.com/SonarSource/sonarlint-visualstudio/blob/c27cd936cb2ff6ebf091d74aab884e1b68ca9bf3/src/IssueViz.Security/Taint/TaintIssuesSynchronizer.cs#L101

Reply:
In both cases, I think, we can rely on the fact that we sync regularly and the missing issue is going to be picked up by the next sync